src/utils/src/ExponentialBackoffUtils.c

```diff
@@ -0,0 +1,211 @@
+
+    pExponentialBackoffState->exponentialBackoffConfig = *pBackoffConfig;
+    CHK_STATUS(resetExponentialBackoffRetryState(pExponentialBackoffState));
+
+CleanUp:
+    if (STATUS_SUCCEEDED(retStatus)) {
+        (*ppBackoffState) = pExponentialBackoffState;
+    } else {
+        SAFE_MEMFREE(pExponentialBackoffState);
+    }
+
+    LEAVES();
+    return retStatus;
+}
+
+UINT64 getRandomJitter(UINT32 jitterFactor) {
+    return (RAND() % jitterFactor) * HUNDREDS_OF_NANOS_IN_A_MILLISECOND;
+}
+
+UINT64 calculateWaitTime(PExponentialBackoffState pRetryState, PExponentialBackoffConfig pRetryConfig) {
+    return power(DEFAULT_KVS_EXPONENTIAL_FACTOR, pRetryState->currentRetryCount) * pRetryConfig->retryFactorTime;
+}
+
+STATUS validateAndUpdateExponentialBackoffStatus(PExponentialBackoffState pExponentialBackoffState) {
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+
+    switch (pExponentialBackoffState->status) {
+        case BACKOFF_NOT_STARTED:
+            pExponentialBackoffState->status = BACKOFF_IN_PROGRESS;
+            DLOGD("Status changed from BACKOFF_NOT_STARTED to BACKOFF_IN_PROGRESS");
+            break;
+        case BACKOFF_IN_PROGRESS:
+            // Do nothing
+            DLOGD("Current status is BACKOFF_IN_PROGRESS");
+            break;
+        case BACKOFF_TERMINATED:
+            DLOGD("Cannot execute exponentialBackoffBlockingWait. Current status is BACKOFF_TERMINATED");
+            CHK(FALSE, STATUS_EXPONENTIAL_BACKOFF_INVALID_STATE);
+            // No 'break' needed since CHK(FALSE, ...) will always jump to CleanUp
+        default:
+            DLOGD("Cannot execute exponentialBackoffBlockingWait. Unexpected state [%"PRIu64"]", pExponentialBackoffState->status);
+            CHK(FALSE, STATUS_EXPONENTIAL_BACKOFF_INVALID_STATE);
+    }
+
+CleanUp:
+    LEAVES();
+    return retStatus;
+}
+
+STATUS exponentialBackoffBlockingWait(PExponentialBackoffState pRetryState) {
+    ENTERS();
+    PExponentialBackoffConfig pRetryConfig = NULL;
+    UINT64 currentTime = 0, waitTime = 0, jitter = 0;
+    STATUS retStatus = STATUS_SUCCESS;
+
+    CHK(pRetryState != NULL, STATUS_NULL_ARG);
+
+    CHK_STATUS(validateAndUpdateExponentialBackoffStatus(pRetryState));
+
+    pRetryConfig = &(pRetryState->exponentialBackoffConfig);
+
+    // If retries is exhausted, return error to the application
+    if (pRetryConfig->maxRetryCount != KVS_INFINITE_EXPONENTIAL_RETRIES) {
+        CHK(!(pRetryConfig->maxRetryCount == pRetryState->currentRetryCount),
+            STATUS_EXPONENTIAL_BACKOFF_RETRIES_EXHAUSTED);
+    }
+
+    // check the last retry time. If the difference between last retry and current time is greater than
+    // minTimeToResetRetryState, then reset the retry state
+    //
+    // Proceed if this is not the first retry
+    currentTime = GETTIME();
+    if (pRetryState->currentRetryCount != 0 && currentTime - pRetryState->lastRetryWaitTime > pRetryConfig->minTimeToResetRetryState) {
```

Comment:
Is it possible that currentTime < pRetryState->lastRetryWaitTime which for unsigned values will result in a wrap around and incorrect behavior?  Should we add a check for that case?

For example there is something similar here:  https://github.com/awslabs/amazon-kinesis-video-streams-pic/blob/master/src/client/src/Client.c#L72

Reply:
done